builder/builder-source-git.c

```diff
@@ -279,70 +279,54 @@ git_mirror_submodules (const char     *repo_location,
+  gsize num_submodules;
 
-  mirror_dir_path = g_file_get_path (mirror_dir);
-
-  checkout_dir_template = g_file_get_child (builder_context_get_state_dir (context),
-                                            "tmp-checkout-XXXXXX");
-  checkout_dir_path = g_file_get_path (checkout_dir_template);
-
-  if (g_mkdtemp (checkout_dir_path) == NULL)
+  if (git (mirror_dir, &submodule_data, NULL, "show", gitmodules, NULL))
     {
-      g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED, "Can't create temporary checkout directory");
-      return FALSE;
-    }
-
-  checkout_dir = g_file_new_for_path (checkout_dir_path);
-
-  if (!git (NULL, NULL, error,
-            "clone", "-q", "--no-checkout", mirror_dir_path, checkout_dir_path, NULL))
-    return FALSE;
-
-  if (!git (checkout_dir, NULL, error, "checkout", "-q", "-f", revision, NULL))
-    return FALSE;
+      if (!g_key_file_load_from_data (key_file, submodule_data, -1,
+                                      G_KEY_FILE_NONE, error))
+        return FALSE;
 
-  if (!git (checkout_dir, &submodule_status, error,
-            "submodule", "status", NULL))
-    return FALSE;
+      submodules = g_key_file_get_groups (key_file, &num_submodules);
 
-  if (submodule_status)
-    {
       int i;
-      g_auto(GStrv) lines = g_strsplit (submodule_status, "\n", -1);
-      for (i = 0; lines[i] != NULL; i++)
+      for (i = 0; i < num_submodules; i++)
         {
-          g_autofree char *url = NULL;
-          g_autofree char *option = NULL;
-          g_autofree char *old = NULL;
+          g_autofree gchar *submodule = NULL;
+          g_autofree gchar *name = NULL;
+          g_autofree gchar *path = NULL;
+          g_autofree gchar *url = NULL;
+          g_autofree gchar *ls_tree = NULL;
           g_auto(GStrv) words = NULL;
-          if (*lines[i] == 0)
+          gsize len;
+
+          submodule = submodules[i];
+          len = strlen (submodule);
+
+          if (!g_str_has_prefix (submodule, "submodule \""))
             continue;
-          words = g_strsplit (lines[i] + 1, " ", 3);
 
-          option = g_strdup_printf ("submodule.%s.url", words[1]);
-          if (!git (checkout_dir, &url, error,
-                    "config", "-f", ".gitmodules", option, NULL))
-            return FALSE;
-          /* Trim trailing whitespace */
-          g_strchomp (url);
+          name = g_strndup (submodule + 11, len - 12);
+          path = g_key_file_get_string (key_file, submodule, "path", error);
+          url = g_key_file_get_string (key_file, submodule, "url", error);
 
-          old = url;
-          url = make_absolute (repo_location, old, error);
           if (url == NULL)
             return FALSE;
 
-          if (!git_mirror_repo (url, update, words[0], context, error))
+          if (!git (mirror_dir, &ls_tree, error, "ls-tree", revision, path, NULL))
+            continue;
+
+          words = g_strsplit_set (ls_tree, " \t", 4);
```

Comment:
Maybe we can add some verification here on the ls_tree output. For instance, make sure its only one line (i.e. it would not if the path was a directory), and at least one line (happens if the path didn't exist). I guess you want to split on \n first. Also, words[1] should be "commit".